pubnub-gson/src/main/java/com/pubnub/internal/endpoints/DelegatingEndpoint.java

```diff
@@ -0,0 +1,75 @@
+package com.pubnub.internal.endpoints;
+
+import com.pubnub.api.PubNubException;
+import com.pubnub.api.endpoints.Endpoint;
+import com.pubnub.api.endpoints.remoteaction.ExtendedRemoteAction;
+import com.pubnub.api.enums.PNOperationType;
+import com.pubnub.api.v2.callbacks.Result;
+import com.pubnub.internal.CorePubNubClient;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.function.Consumer;
+
+public abstract class DelegatingEndpoint<T> implements Endpoint<T> {
```

Comment:
This is general comment. I think it would be valuable to create doc presenting new architecture or some main concepts. Current design is complicated IMHO e.g. we have 7 types(interfaces, classes) that are responsible for representing PubNub object
BasePubNub
BasePubNubImpl
PubNub.kt
PubNub.java
PubNubImpl.kt
PubNubImpl.java
corePubNubClient.kt

For particular operation we have 6 types e.g.
DeleteMessages.kt
DeleteMessages.java
DeleteMessagesClass.kt
DeleteMessagesImpl.java
DeleteMessagesImpl.kt
IdeleteMessages.kt

Reply:
agreed